### PR TITLE
tests: Bluetooth: Mesh: Add tests to cover message publication

### PR DESCRIPTION
**Pull request by PavelVPV**
This PR adds tests that cover message publication behavior in the access layer and fixes publication period measurement in the access layer.

The legacy advertiser was replaced with the extended one. I see more messages in the scan callback on the receiver that expected to be sent be the sender even if the network transmit count state is 0. The extended advertiser seems to be working correctly. This has to be investigated separately.

### CODE REVIEW

File: subsys/bluetooth/mesh/access.c [hunk 1]
```diff
@@ -259,12 +252,14 @@ static void mod_publish(struct k_work *work)
 	}
 
 	err = publish_transmit(pub->mod);
+
+	/* Initialize the timestamp for the beginning of a new period */
+	if (pub->count == BT_MESH_PUB_TRANSMIT_COUNT(pub->retransmit)) {
+		pub->period_start = k_uptime_get_32();
+	}
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/access.c:259]: Should be moved into `pub_period_start()` so we can ditch this count logic

File: subsys/bluetooth/mesh/access.c [hunk 2]
```diff
@@ -116,9 +116,14 @@ static int32_t next_period(struct bt_mesh_model *mod)
 	struct bt_mesh_model_pub *pub = mod->pub;
 	uint32_t elapsed, period;
 
-	period = bt_mesh_model_pub_period_get(mod);
-	if (!period) {
-		return 0;
+	if (mod->pub->count) {
+		period = BT_MESH_PUB_TRANSMIT_INT(mod->pub->retransmit);
+		period *= BT_MESH_PUB_MSG_NUM(mod->pub);
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/access.c:121]: This logic could use a comment, I think.

File: subsys/bluetooth/mesh/access.c [hunk 3]
```diff
@@ -116,9 +116,14 @@ static int32_t next_period(struct bt_mesh_model *mod)
...
+	} else {
+		period = bt_mesh_model_pub_period_get(mod);
+		if (!period) {
+			return 0;
+		}
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/access.c:126]: Should perform this check in the first branch too, just to make sure the handling of changing retransmit counts is coherent. In that branch, we should probably fall back to this branch though, so we can schedule the next period if the count has changed.

File: tests/bluetooth/bsim_bt/bsim_test_mesh/prj.conf [hunk 4]
```diff
@@ -44,3 +44,9 @@ CONFIG_BT_MESH_PROV_OOB_PUBLIC_KEY=y
 CONFIG_BT_MESH_MODEL_EXTENSIONS=y
 
 CONFIG_BT_MESH_DEBUG=y
+
+# Replace the legacy advertiser with the extended one due to a bug in simulation
+CONFIG_BT_CTLR_ADV_EXT=y
+CONFIG_BT_EXT_ADV=y
+CONFIG_BT_EXT_ADV_MAX_ADV_SET=1
+CONFIG_BT_MESH_ADV_EXT=y
```
Comment (trond-snekvik) [inline tests/bluetooth/bsim_bt/bsim_test_mesh/prj.conf:52]: > When using the legacy advertiser, I see more messages in the scan
callback on a receiver device than were supposed to be sent out from a
sender device even if the network transmit count state is zero.

This is a side effect of the inaccuracy of the legacy advertiser mechanism, and can't be fixed. Could be fixed in the test by ignoring duplicate advertisements, I think

File: subsys/bluetooth/mesh/access.c [hunk 5]
```diff
@@ -116,9 +116,14 @@ static int32_t next_period(struct bt_mesh_model *mod)
...
 	}
 
 	elapsed = k_uptime_get_32() - pub->period_start;
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/access.c:129]: The warning log on line 134 will probably start showing up a lot more now, and might need more accurate text

**Review by trond-snekvik — CHANGES_REQUESTED** [review]

File: tests/bluetooth/bsim_bt/bsim_test_mesh/prj.conf (hunk 4 above)
Comment (PavelVPV) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/prj.conf:52]: They can't be ignored at the top and affects on test's measurements. I will have to increase jitter margin on a receiver.

Comment (trond-snekvik) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/prj.conf:52]: Wouldn't it be possible to ignore raw byte-for-byte duplicates? Both access and transport repeats will have different seqnums

File: tests/bluetooth/bsim_bt/bsim_test_mesh/prj.conf [hunk 6]
```diff
@@ -1,10 +1,10 @@
-CONFIG_BT=y
-CONFIG_BT_DEBUG_LOG=y
 CONFIG_LOG_MODE_IMMEDIATE=y
 CONFIG_ASSERT=y
+CONFIG_SYS_CLOCK_TICKS_PER_SEC=1000
 
 # Bluetooth configuration
 CONFIG_BT=y
+CONFIG_BT_DEBUG_LOG=y
```
[the remark this answers is not in the document]
Comment (PavelVPV) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/prj.conf:7]: I have no opinion on this. This worked fine before. Probably ok to leave it for now?

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_access.c [hunk 7]
```diff
@@ -245,6 +325,335 @@ static void test_sub_ext_model(void)
...
+		int32_t pub_delta = llabs(time_delta - interval);
+
+		jitter = MAX(pub_delta, jitter);
+
+		LOG_DBG("Recv time: %d delta: %d jitter: %d", (int32_t)timestamp, time_delta,
+			jitter);
+	}
+
+	LOG_INF("Recv jitter: %d", jitter);
+	ASSERT_TRUE(jitter <= 35);
```
Comment (alxelax) [inline tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_access.c:419]: How did you choose the jitter values for comparisons? Will be jitter worse on PCs with low performance?

Comment (PavelVPV) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_access.c:419]: Roughly it should not be bigger than the advertising event time multiplied by number of network transmissions. If there is not packet loss, it should be received within the first advertising event. I had to increase it a bit due to a packet loss.

I don't think it should depend on PC performance.

File: tests/bluetooth/bsim_bt/bsim_test_mesh/prj.conf [hunk 8]
```diff
@@ -44,3 +44,9 @@ CONFIG_BT_MESH_PROV_OOB_PUBLIC_KEY=y
 CONFIG_BT_MESH_MODEL_EXTENSIONS=y
 
 CONFIG_BT_MESH_DEBUG=y
+
+# Replace the legacy advertiser with the extended one due to a bug in simulation
```
[the remark this answers is not in the document]
Comment (PavelVPV) [reply tests/bluetooth/bsim_bt/bsim_test_mesh/prj.conf]: I removed this change.

File: tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_access.c [hunk 9]
```diff
@@ -39,6 +44,82 @@ static uint8_t app_key[16] = { 0xaa };
...
+};
+
+/* Test vector for canceling a message publication. */
+static struct {
+	uint8_t period;
+	uint8_t transmit;
+	uint8_t msgs;
+	int32_t sleep;
+	int32_t duration;
+} test_cancel[] = {
```
Comment (LingaoM) [inline tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_access.c:74]: I always pay attention to whether const is required first, but after thinking about it, in fact, running the script on a Linux device doesn't care about these few bytes.

**Review by alxelax — APPROVED** [review]

**Review by trond-snekvik — APPROVED** [review]

**Review by LingaoM — APPROVED** [review]

**Review by Balaklaka — APPROVED** [review]

**Pull request merged**